airbyte-cdk/python/airbyte_cdk/sources/file_based/config/file_based_stream_config.py

```diff
@@ -0,0 +1,76 @@
+#
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+#
+
+import codecs
+from typing import Any, List, Mapping, Optional, Union
+
+from airbyte_cdk.models import ConfiguredAirbyteCatalog
+from pydantic import BaseModel, root_validator, validator
+
+PrimaryKeyType = Optional[Union[str, List[str], List[List[str]]]]
+
+
+class CsvFormat(BaseModel):
+    filetype: str
+    delimiter: str = ","
+    quote_char: str = '"'
+    escape_char: Optional[str]
+    encoding: Optional[str] = "utf8"
+    double_quote: bool
+    # Noting that the existing S3 connector had a config option newlines_in_values. This was only supported by pyarrow and not
+    # the Python csv package. It has a little adoption, but long term we should ideally phase this out because of the drawbacks
+    # of using pyarrow
+
+    @validator("delimiter")
+    def validate_delimiter(cls, v):
+        if len(v) != 1:
+            raise ValueError("delimiter should only be one character")
+        if v in {"\r", "\n"}:
+            raise ValueError(f"delimiter cannot be {v}")
+        return v
+
+    @validator("quote_char")
+    def validate_quote_char(cls, v):
+        if len(v) != 1:
+            raise ValueError("quote_char should only be one character")
+        return v
+
+    @validator("escape_char")
+    def validate_escape_char(cls, v):
+        if len(v) != 1:
+            raise ValueError("escape_char should only be one character")
+        return v
+
+    @validator("encoding")
+    def validate_encoding(cls, v):
+        try:
+            codecs.lookup(v)
+        except LookupError:
+            raise ValueError(f"invalid encoding format: {v}")
+        return v
+
+
+class FileBasedStreamConfig(BaseModel):
+    name: str
+    file_type: str
+    globs: Optional[List[str]]
+    validation_policy: str
+    catalog_schema: Optional[ConfiguredAirbyteCatalog]
+    input_schema: Optional[Mapping[str, Any]]
+    primary_key: PrimaryKeyType
+    max_history_size: Optional[int]
+    days_to_sync_if_history_is_full: Optional[int]
+    format: Optional[CsvFormat]  # this will eventually be a Union once we have more than one format type
```

Comment:
i'm not sure about enable? but i based this on the existing S3 implementation. The `format` object is built when you expand the optional block in the connector and then whatever you specify will be included in this object. Its existence tells us to build the dialect

Reply:
yeah that's a good point it might set us up better for that in the future. But we are still specifying `file_type` in the config though so we're not quite at the point where we can handle multiple ones yet. I think I can make this work, but I think this would potentially impact how we would migrate existing S3 connections since format is correlated to one object. 

But simplify things and not need to worry about a migration in the immediate, how would you feel if we added a bit of logic when we parse the config into the pydantic model. If the `format` is an object that contains `filetype` (the existing format), we instead transform the mapping to `csv` -> `format` object?

Reply:
👍 I like that idea.